tools/traffic_vault_migrate/postgres.go

```diff
@@ -0,0 +1,743 @@
+		}
+		i += 1
+	}
+	return nil
+}
+func (tbl *pgDNSSecTable) decrypt(aesKey []byte) error {
+	for i, _ := range tbl.Records {
+		err := decryptInto(aesKey, tbl.Records[i].DataEncrypted, &tbl.Records[i].Key)
+		if err != nil {
+			return fmt.Errorf("unable to decrypt into keys: %w", err)
+		}
+	}
+	return nil
+}
+func (tbl *pgDNSSecTable) encrypt(aesKey []byte) error {
+	for i, dns := range tbl.Records {
+		data, err := json.Marshal(&dns.Key)
+		if err != nil {
+			return fmt.Errorf("encrypt issue marshalling keys: %w", err)
+		}
+		dat, err := encrypt(data, aesKey)
+		if err != nil {
+			return fmt.Errorf("encrypt error: %w", err)
+		}
+		tbl.Records[i].DataEncrypted = dat
+	}
+	return nil
+}
+func (tbl *pgDNSSecTable) toGeneric() []DNSSecKey {
+	keys := make([]DNSSecKey, len(tbl.Records))
+
+	for i, record := range tbl.Records {
+		keys[i] = DNSSecKey{
+			CDN:                    record.CDN,
+			DNSSECKeysTrafficVault: record.Key,
+		}
+	}
+
+	return keys
+}
+func (tbl *pgDNSSecTable) fromGeneric(keys []DNSSecKey) {
+	tbl.Records = make([]pgDNSSecRecord, len(keys))
+
+	for i, key := range keys {
+		tbl.Records[i] = pgDNSSecRecord{
+			Key: key.DNSSECKeysTrafficVault,
+			CDN: key.CDN,
+			pgCommonRecord: pgCommonRecord{
+				DataEncrypted: nil,
+			},
+		}
+	}
+}
+func (tbl *pgDNSSecTable) validate() []string {
+	for i, record := range tbl.Records {
+		if record.DataEncrypted == nil && len(record.Key) > 0 {
+			return []string{fmt.Sprintf("DNSSEC Key %v: DataEncrypted is blank!", i)}
+		}
+	}
+	return nil
+}
+func (tbl *pgDNSSecTable) insertKeys(db *sql.DB) error {
+	queryBase := "INSERT INTO dnssec (cdn, data) VALUES "
+	stride := 2
+	queryArgs := make([]interface{}, len(tbl.Records)*stride)
+	for i, record := range tbl.Records {
+		j := i * stride
+		queryArgs[j] = record.CDN
+		queryArgs[j+1] = record.DataEncrypted
+	}
+	return insertIntoTable(db, queryBase, stride, queryArgs)
+}
+
+type pgSSLKeyRecord struct {
+	Keys tc.DeliveryServiceSSLKeys
+	pgCommonRecord
+
+	// These records are stored on the table but are duplicated
+	DeliveryService string
+	CDN             string
+}
+type pgSSLKeyTable struct {
+	Records []pgSSLKeyRecord
+}
+
+func (tbl *pgSSLKeyTable) insertKeys(db *sql.DB) error {
+	queryBase := "INSERT INTO sslkey (deliveryservice, data, cdn, version) VALUES "
+	duplicateKeys := 2
+	stride := 4
+	queryArgs := make([]interface{}, len(tbl.Records)*stride*duplicateKeys)
+	for i, record := range tbl.Records {
+		j := i * duplicateKeys * stride
+
+		queryArgs[j] = record.DeliveryService
+		queryArgs[j+1] = record.DataEncrypted
+		queryArgs[j+2] = record.CDN
+		queryArgs[j+3] = record.Keys.Version.String()
+
+		queryArgs[j+4] = record.DeliveryService
+		queryArgs[j+5] = record.DataEncrypted
+		queryArgs[j+6] = record.CDN
+		queryArgs[j+7] = "latest"
+	}
+	return insertIntoTable(db, queryBase, 4, queryArgs)
+}
+func (tbl *pgSSLKeyTable) gatherKeys(db *sql.DB) error {
+	sz, err := getSize(db, "sslkey WHERE version='latest'")
```

Comment:
The other counterpart to each latest record is identical, it is inserted twice for searching/indexing.The reason I chose to only get latest records was that otherwise there are duplicate `SSLKey` records.

Reply:
Right, if we only care to get the latest version of an sslkey, then getting `latest` is fine. But I think we might want to migrate all _prior_ versions too.

For instance, let's say a DS is on ssl_key_version 5. So it currently has all these sslkey versions stored:
- 0
- 1
- 2
- 3
- 4
- 5
- latest

Versions `5` and `latest` are the same sslkey, but we might also want to migrate versions `0` through `4`.

Reply:
Hmmm...I was misreading how we inserted ssl keys, you are correct. I'll remove the filter.